### PR TITLE
Expose node status so that external preemption plugins can use it

### PR DESCRIPTION
**Pull request by chendave**
**What type of PR is this?**

Add one of the following kinds:
/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
This is following up the comments from here: https://github.com/kubernetes/kubernetes/pull/97184#discussion_r553991578
**Does this PR introduce a user-facing change?**:

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go [hunk 1]
```diff
@@ -265,18 +266,21 @@ func PodEligibleToPreemptOthers(pod *v1.Pod, nodeInfos framework.NodeInfoLister,
...
-func nodesWherePreemptionMightHelp(nodes []*framework.NodeInfo, m framework.NodeToStatusMap) []*framework.NodeInfo {
+func nodesWherePreemptionMightHelp(nodes []*framework.NodeInfo, m framework.NodeToStatusMap) ([]*framework.NodeInfo, framework.NodeToStatusMap) {
 	var potentialNodes []*framework.NodeInfo
+	nodeStatuses := make(framework.NodeToStatusMap)
 	for _, node := range nodes {
 		name := node.Node().Name
 		// We reply on the status by each plugin - 'Unschedulable' or 'UnschedulableAndUnresolvable'
 		// to determine whether preemption may help or not on the node.
 		if m[name].Code() == framework.UnschedulableAndUnresolvable {
+			message := fmt.Sprintf("Preemption is not helpful for scheduling")
```
Comment (adtac) [inline pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:277]: the fmt.Sprintf call isn't required here

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go [hunk 2]
```diff
@@ -220,15 +226,10 @@ func (pl *DefaultPreemption) FindCandidates(ctx context.Context, state *framewor
...
 	candidates, nodeStatuses := dryRunPreemption(ctx, pl.fh, state, pod, potentialNodes, pdbs, offset, numCandidates)
-	// Return a FitError only when there are no candidates that fit the pod.
-	if len(candidates) == 0 {
-		return candidates, &framework.FitError{
-			Pod:                   pod,
-			NumAllNodes:           len(potentialNodes),
-			FilteredNodesStatuses: nodeStatuses,
-		}
+	for node, status := range unschedulableNodeStatus {
+		nodeStatuses[node] = status
```
Comment (adtac) [inline pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:230]: what happens if a node is a part of both maps? should we overwrite the dryRunPreemption status with the unschedulableNodeStatus value?

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go (hunk 1 above)
Comment (chendave) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:277]: yes, will update soon.

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go (hunk 2 above)
Comment (chendave) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:230]: Not sure if I understand your question correctly, the `unschedulableNodeStatus` we got here are from the nodes that are have not evaluated by `dryRunPreemption`. 

`len(unschedulableNodeStatus) + len(potentialNodes) = allNodes`

I cannnot see where the overwrite could happen.

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go (hunk 1 above)
Comment (chendave) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go]: Done

**Comment by chendave** [comment]
@adtac thanks for the review!

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go [hunk 3]
```diff
@@ -186,29 +193,28 @@ func (pl *DefaultPreemption) getOffsetAndNumCandidates(numNodes int32) (int32, i
...
+	potentialNodes, unschedulableNodeStatus := nodesWherePreemptionMightHelp(allNodes, m)
 	if len(potentialNodes) == 0 {
 		klog.V(3).Infof("Preemption will not help schedule pod %v/%v on any node.", pod.Namespace, pod.Name)
 		// In this case, we should clean-up any existing nominated node name of the pod.
 		if err := util.ClearNominatedNodeName(pl.fh.ClientSet(), pod); err != nil {
 			klog.Errorf("Cannot clear 'NominatedNodeName' field of pod %v/%v: %v", pod.Namespace, pod.Name, err)
 			// We do not return as this error is not critical.
 		}
-		return nil, nil
+		return nil, len(allNodes), unschedulableNodeStatus, nil
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:212]: Do we need the length of all nodes? as all other PostFilter plugins also know that.

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go [hunk 4]
```diff
@@ -138,10 +138,17 @@ func (pl *DefaultPreemption) preempt(ctx context.Context, state *framework.Cycle
...
 
 	// 2) Find all preemption candidates.
-	candidates, err := pl.FindCandidates(ctx, state, pod, m)
-	if err != nil || len(candidates) == 0 {
+	candidates, nodeNum, status, err := pl.FindCandidates(ctx, state, pod, m)
+	if err != nil {
 		return "", err
 	}
+	if len(candidates) == 0 {
+		return "", &framework.FitError{
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:146]: It's not clear to me that how this returned obj is utilized by subsequent PostFilter plugins.

Comment (chendave) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:146]: This PR intends to change the public method of `FindCandidates` to return node `status`, so that the calling of this method could get the node status as well.

Quote from @ahg-g 's initial comments, 
> This is just so external preemption plugins that use this interface can benefit from this info, but this is less critical for this PR, so I guess we can leave this here for now.


As to the return `FitError` here, it's used to populate the log message on the preemption failure.  I did some refactoring to move this obj from  line 224 to here, the reason is I need to add the `unschedulableNodeStatus` into the status or else user might confuse that they have for example they have 3 nodes in the cluster, but only give hint on one node that doesn't have enough resource but no idea on what's happening on other two nodes (please see my update on the testcases).

In short, it is sill used by log population.

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go (hunk 3 above)
Comment (chendave) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:212]: yes, I need to build a obj of `FitError` where how many node are evaluated are needed there.

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go (hunk 4 above)
Comment (Huang-Wei) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:146]: > so that the calling of this method could get the node status as well.

Well, the current version is not stateless so arguably I doubt if users would like to use it... But I see your point.

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go [hunk 5]
```diff
@@ -138,10 +138,17 @@ func (pl *DefaultPreemption) preempt(ctx context.Context, state *framework.Cycle
 	}
 
 	// 2) Find all preemption candidates.
-	candidates, err := pl.FindCandidates(ctx, state, pod, m)
-	if err != nil || len(candidates) == 0 {
+	candidates, nodeNum, status, err := pl.FindCandidates(ctx, state, pod, m)
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:141]:
```suggestion
	candidates, nodesNum, nodeToStautsMap, err := pl.FindCandidates(ctx, state, pod, m)
```

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go [hunk 6]
```diff
@@ -186,29 +193,28 @@ func (pl *DefaultPreemption) getOffsetAndNumCandidates(numNodes int32) (int32, i
 
 // FindCandidates calculates a slice of preemption candidates.
 // Each candidate is executable to make the given <pod> schedulable.
-func (pl *DefaultPreemption) FindCandidates(ctx context.Context, state *framework.CycleState, pod *v1.Pod, m framework.NodeToStatusMap) ([]Candidate, error) {
+func (pl *DefaultPreemption) FindCandidates(ctx context.Context, state *framework.CycleState, pod *v1.Pod, m framework.NodeToStatusMap) ([]Candidate, int, framework.NodeToStatusMap, error) {
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:196]: It seems in all return statements, the 2nd return parameter always equals to len(NodeToStatusMap), can you confirm? If so, we can remove the 2nd return param.

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go [hunk 7]
```diff
@@ -265,18 +266,20 @@ func PodEligibleToPreemptOthers(pod *v1.Pod, nodeInfos framework.NodeInfoLister,
 
 // nodesWherePreemptionMightHelp returns a list of nodes with failed predicates
 // that may be satisfied by removing pods from the node.
-func nodesWherePreemptionMightHelp(nodes []*framework.NodeInfo, m framework.NodeToStatusMap) []*framework.NodeInfo {
+func nodesWherePreemptionMightHelp(nodes []*framework.NodeInfo, m framework.NodeToStatusMap) ([]*framework.NodeInfo, framework.NodeToStatusMap) {
 	var potentialNodes []*framework.NodeInfo
+	nodeStatuses := make(framework.NodeToStatusMap)
 	for _, node := range nodes {
 		name := node.Node().Name
 		// We reply on the status by each plugin - 'Unschedulable' or 'UnschedulableAndUnresolvable'
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go:274]: Can you help fix the typo here? reply -> rely

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go [hunk 8]
```diff
@@ -154,7 +154,7 @@ func TestPostFilter(t *testing.T) {
 				"node1": framework.NewStatus(framework.UnschedulableAndUnresolvable),
 			},
 			wantResult: nil,
-			wantStatus: framework.NewStatus(framework.Unschedulable),
+			wantStatus: framework.NewStatus(framework.Unschedulable, "0/1 nodes are available: 1 Preemption is not helpful for scheduling."),
```
Comment (Huang-Wei) [inline pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:157]: Can you add a test like: 2 nodes failed on nodesWherePreemptionMightHelp() (due to UnschedulableAndUnresolvable status), and 2 nodes failed in dryRunPreemption()?

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go (hunk 5 above)
Comment (chendave) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go]: Done

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go (hunk 6 above)
Comment (chendave) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go]: Great suggestion! they should be equal

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go (hunk 7 above)
Comment (chendave) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go]: Done

File: pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go (hunk 8 above)
Comment (chendave) [reply pkg/scheduler/framework/plugins/defaultpreemption/default_preemption_test.go:157]: Done

**Comment by chendave** [comment]
rebased and comments has been addressed.
@Huang-Wei  could you please take another look?

**Comment by Huang-Wei** [comment]
@chendave LGTM. Could you squash the commits?

**Comment by chendave** [comment]
> @chendave LGTM. Could you squash the commits?

Squashed commits into 2.
@Huang-Wei , I feel we can keep two commits in this PR, 1) Expose node status 2) Add `UnschedulableAndUnresolvable` node into final node status.  
please let me know your thoughts?
